File: cfassist-go/internal/context/context.go
package context

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

var supportedExtensions = map[string]bool{
	".md": true, ".txt": true, ".yaml": true, ".yml": true,
	".csv": true, ".json": true, ".conf": true, ".cfg": true,
	".ini": true, ".toml": true,
}

// LoadDirectory reads all context files from a directory.
// Returns the combined text and file count.
func LoadDirectory(directory string, maxChars int) (string, int) {
	if maxChars <= 0 {
		maxChars = 32000
	}

	info, err := os.Stat(directory)
	if err != nil || !info.IsDir() {
		return "", 0
	}

	// Find all supported files
	var files []string
	filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return nil
		}
		ext := strings.ToLower(filepath.Ext(path))
		if supportedExtensions[ext] {
			files = append(files, path)
		}
		return nil
	})

	sort.Strings(files)

	var parts []string
	charCount := 0
	fileCount := 0

	for _, f := range files {
		data, err := os.ReadFile(f)
		if err != nil {
			continue
		}

		relPath, _ := filepath.Rel(directory, f)
		header := fmt.Sprintf("--- File: %s ---", relPath)
		content := string(data)
		entry := header + "\n" + content + "\n"

		if charCount+len(entry) > maxChars {
			// Try to fit a truncated version
			remaining := maxChars - charCount - len(header) - 20 // room for header + marker
			if remaining > 100 {
				entry = header + "\n" + content[:remaining] + "\n[truncated]\n"
				parts = append(parts, entry)
				fileCount++
			}
			break
		}

		parts = append(parts, entry)
		charCount += len(entry)
		fileCount++
	}

	return strings.Join(parts, "\n"), fileCount
}

File: cfassist/cfassist/context.py
"""Load context files from a directory into the system prompt."""

from pathlib import Path

SUPPORTED_EXTENSIONS = {".md", ".txt", ".yaml", ".yml", ".csv", ".json", ".conf", ".cfg", ".ini", ".toml"}


def load_context_directory(directory, max_chars=32000):
    """Load all supported files from directory into a single text block.

    Files are sorted alphabetically (use 01-, 02- prefixes for ordering).
    Each file is separated with a header showing the filename.

    Returns:
        (text, file_count) tuple
    """
    dir_path = Path(directory)
    if not dir_path.exists():
        return None, 0

    files = sorted(
        f for f in dir_path.rglob("*")
        if f.is_file() and f.suffix.lower() in SUPPORTED_EXTENSIONS
    )

    if not files:
        return None, 0

    parts = []
    total_chars = 0

    for file_path in files:
        try:
            content = file_path.read_text(errors="replace")
        except Exception:
            continue

        # Relative path from context directory for the header
        relative = file_path.relative_to(dir_path)
        header = f"--- File: {relative} ---"
        entry = f"{header}\n{content}\n"

        if total_chars + len(entry) > max_chars:
            # Add what fits, then stop
            remaining = max_chars - total_chars
            if remaining > len(header) + 100:  # only add if meaningful content fits
                parts.append(f"{header}\n{content[:remaining - len(header) - 10]}\n[truncated]\n")
            break

        parts.append(entry)
        total_chars += len(entry)

    text = "\n".join(parts)
    return text, len(parts)
